### PR TITLE
idl-doc: Further improvements

### PR DESCRIPTION
**Pull request by apelisse**

### CODE REVIEW

**Comment by apelisse** [comment]
I've actually added much more changes.

File: pkg/idl/doc.go [hunk 1]
```diff
@@ -14,49 +14,126 @@ See the License for the specific language governing permissions and
...
-// Using that tag will generate the following OpenAPI extension:
+// Using this tag will generate the following OpenAPI extension:
 //  "x-kubernetes-list-type": "set"
 type ListType string
+
+// ListMapKey annotates map lists by specifying the key used as the index of the map.
+//
+// This tag MUST only be used on lists that have the listType=map
+// attribute, or the generation step will fail. Also, the value
+// specified for this attribute must be a field, of a scalar type, of
```
Comment (pwittrock) [inline pkg/idl/doc.go]: > field, of a scalar type

scalar typed field

File: pkg/idl/doc.go [hunk 2]
```diff
@@ -14,49 +14,126 @@ See the License for the specific language governing permissions and
...
+//  "x-kubernetes-map-type": "atomic"
+type MapType string
+
+// OpenAPIGen needs to be described.
+type OpenAPIGen string
+
+// Optional needs to be described.
+type Optional string
+
+// PatchStrategy needs to be described.
```
Comment (pwittrock) [inline pkg/idl/doc.go]: These will be deprecated I think

**Comment by pwittrock** [comment]
lgtm

File: pkg/idl/doc.go (hunk 2 above)
Comment (apelisse) [reply pkg/idl/doc.go]: Yeah I still want to describe them as such

**Comment by apelisse** [comment]
Addressed the comment and fixed a couple of other bugs that I noticed.

File: pkg/idl/doc.go [hunk 3]
```diff
@@ -14,49 +14,126 @@ See the License for the specific language governing permissions and
...
 // Set
 //
-// The tag looks like this:
+// Example:
 //  +listType=set
 //
-// Sets are lists that can't have multiple times the same value. Each
-// value must be a scalar or an atomic list and can't be map, set or struct.
+// Sets are lists that must not have multiple times the same value. Each
+// value must be a scalar.
```
Comment (lavalamp) [inline pkg/idl/doc.go]: a scalar (or another atomic type).

File: pkg/idl/listtype_test.go [hunk 4]
```diff
@@ -16,40 +16,41 @@ limitations under the License.
...
+	}
 	type SomeAPI struct {
 		// +listType=map
 		// +listMapKey=name
-		elements []struct {
-			Name string
-
-			Value int
-		}
+		elements []SomeStruct
```
Comment (lavalamp) [inline pkg/idl/listtype_test.go]: I would like more clarity that the tags are there on account of the list type and not on account of the field. Next PR?

File: pkg/idl/doc.go [hunk 5]
```diff
@@ -14,49 +14,126 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-/*
-The IDL package describes comment directives that may be applied to API types.
-*/
+// The IDL package describes comment directives that may be applied to API types.
```
Comment (lavalamp) [inline pkg/idl/doc.go]: Types and fields. And we should say which tags are which.

**Comment by lavalamp** [comment]
LGTM, can you squash?

File: pkg/idl/listtype_test.go (hunk 4 above)
Comment (apelisse) [reply pkg/idl/listtype_test.go]: Let's do next PR so that we don't mix too many changes together.

**Comment by apelisse** [comment]
Fixed and squashed! Thanks Daniel, please merge when tests pass.

**Pull request merged**